design/design.md

```diff
@@ -0,0 +1,223 @@
+# Native support for GRPC in rippled
+
+## Overview
+GRPC is an RPC system that allows client applications to directly call methods
+on a server application as if it was a local object. gRPC uses protocol buffers
+as its Interface Definition Language, as well as the underlying message
+interchange format. gRPC has bindings in many languages. gRPC uses the protocol
+buffer compiler (`protoc`) with the gRPC plugin to generate data access classes
+as well as client and server code. This generated code handles serialization,
+deserialization and I/O. The messages and methods are defined in `.proto` files
+that are used by both client and server. For more information about gRPC and
+protocol buffers, go [here](https://grpc.io/docs/).
+
+Currently, rippled supports RPC over Web Socket, REST-JSON and CLI. gRPC would
+be in addition to these existing RPC technologies. The reason for adding gRPC
+support to rippled is that gRPC is very easy to use as a client, and has
+bindings in many languages, allowing developers to more quickly develop
+applications without worrying about small details and nuances related to RPC.
+
+## ExecutionConcept
+
+In `Application::setup()`, a grpc server begins listening for requests,
+and an event loop begins running in a separate, dedicated
+thread; this thread is used exclusively for servicing grpc requests, and is an
+additional thread used by the application, separate from the threads used by the
+JobQueue or the I/O threads used by asio. In the setup of the event loop, each rpc 
+call (request type), is bound to the `CompletionQueue`; when a request arrives,
+the request is placed on the `CompletionQueue`. The event loop is an infinite
+loop that queries the `CompletionQueue` by calling `CompletionQueue::Next()`,
+which returns the next event that has occurred, blocking until such an event is
+available. The events, in our application, can be one of two things:
+* A request has arrived.
+* A response has been sent.
+
+When `CompletionQueue::Next()` returns a request, the event loop posts a
+coroutine onto the JobQueue, to be invoked later; this coroutine, when invoked, 
+calls the corresponding handler, populates the response, and submits the response for sending.
+Posting to the JobQueue in this way frees up the event loop to process more events
+while requests are being handled. Once grpc sends the response, an event is
+placed on the `CompletionQueue`; when `CompletionQueue::Next()` returns such an
+event, signaling the response has been sent, the event loop cleans up the resources
+associated with this request/response.
```

Comment:
Mark Travis had questions on how we support concurrent sessions asynchronously. We achieve this by using CallData. I know that you have some description down below but i would like to expand on this topic:
mainly I want to see how exactly we achieve the concurrent session support at the implementation level described in words. (dynamically creating calldata). and how the objects created are managed/exeucted/destructed. I think you touched on them via doCreate, doProcess and doFinish, but to me im still unclear on the flow. (im leaving this comment after i have read the entire doc).